### PR TITLE
Add retriever

### PR DESCRIPTION
**Pull request by wpears**
Combines the loader and retriever into a single project. Major features:

  - Cli calls retriever which calls the loader
    - retriever just pulls in data and transforms it to JSON
    - loader prepares for bulk upload and loads into elasticsearch, handling aliasing/deleting old aliases
  - S3 is now only invoked for data backups
    - retriever's JSON output is converted to CSV, gzipped and uploaded.
    - backing up data is optional
    - Ability to fall back to previously backed up data is coming in a later PR
  - Tiger support will be added in a different PR, see #112

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+if(esVar){
+  esVar = esVar.split('//')[1].split(':');
+  esHost = esVar[0];
+  esPort = +esVar[1];
+}
+options
+  .version('0.0.1')
+  .option('-f --file <file>', 'The json data file that contains the collected data endpoints and field mappings.')
+  .option('-m --match <match>', 'A string or regular expression that the names from the <file> must contain or match')
+  .option('-h, --host <host>', 'ElasticSearch host. Defaults to localhost unless linked to a Docker container aliased to Elasticsearch', esHost || 'localhost')
```
Comment (hkeeler) [inline index.js]: Some `option`s have a comma separating short and long option, but not others.  Does this have any significance?  If no, probably want to make that consistent.

Comment (wpears) [inline index.js]: Updated

File: index.js [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+  //loader(options, stream, callback
+  //
+  logger.info('%d error%s encountered.',
+    output.errors.length,
+    output.errors.length === 1 ? '' : 's'
+  );
+
+  output.errors.forEach(function(v, i){
+    logger.error(v);
+    output.errors[i] = v.toString();
```
Comment (hkeeler) [inline index.js]: Why is this necessary?  Why reset the list to a set of `String`s instead of `Error`s?

File: index.js [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+  }else{
+    logger.info('Fetched %d source%s and placed %s in %s.',
+      output.retrieved.length,
+      output.retrieved.length === 1 ? '' : 's',
+      output.retrieved.length === 1 ? 'it' : 'them',
+      output.location
+    );
+  }
+
+  console.log(JSON.stringify(output));
```
Comment (hkeeler) [inline index.js]: Why `console.log` instead using of `logger`?

Comment (wpears) [inline index.js]: Good question! Just an artifact of the old loader. Will fix.

File: index.js (hunk 2 above)
Comment (wpears) [inline index.js]: errors don't stringify; so this is done to support logging the stringified output

File: lib/bulkPrefixer.js [hunk 4]
```diff
@@ -0,0 +1,10 @@
+'use strict';
```
Comment (hkeeler) [inline lib/bulkPrefixer.js]: I'm not quite clear on the intent this function.  Could you add a comment as to what this does and why we need it?

Comment (wpears) [inline lib/bulkPrefixer.js]: This enables [strict mode](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Strict_mode) which prevents some weird behavior from older versions of javascript

Comment (hkeeler) [inline lib/bulkPrefixer.js]: Sorry, no I meant the `bulkPrefixer` in general.  I see that its concat-ing the index JSON onto the chunk, but what's the purpose?

Comment (wpears) [inline lib/bulkPrefixer.js]: Ahah! the elasticsearch bulk API allows you to load many documents with one request, to do so you need to prefix the document with bulk loading information (in this case, just that we want to index the document)

Comment (hkeeler) [inline lib/bulkPrefixer.js]: Ah.  OK.  Makes sense.

Also, could you jump on the group chat.  There's much loader chat going on about how to load the TIGER data.

**Comment by hkeeler** [comment]
Looks good.  Definitely easier to follow with the merger of retriever and all of @wpears's cleanup.  I'm am going to merge this, but this was a _very_ large PR, and it would be very helpful if others (@jmarin @awolfe76) could jump in and take a peek too.
